src/main/scala-2.11/org/apache/james/gatling/smtp/SmtpAction.scala

```diff
@@ -1,70 +1,54 @@
 import io.gatling.core.session.Session
 import io.gatling.core.stats.StatsEngine
-import io.gatling.core.stats.message.ResponseTimings
-import org.apache.commons.mail.SimpleEmail
 import org.apache.james.gatling.control.UserFeeder
 
-import scala.util.{Failure, Success, Try}
+import scala.util.Failure
 
-class SmtpAction(
-                  requestName: String,
+object SmtpAction {
+  def props(requestName: String, subject: String, body: String, statsEngine: StatsEngine, next: Action, protocol: SmtpProtocol) =
+    Props(new SmtpAction(requestName, subject, body, statsEngine, next, protocol))
+}
+
+class SmtpAction(requestName: String,
                   subject: String,
                   body: String,
                   val statsEngine: StatsEngine,
                   val next: Action,
-                  protocol: SmtpProtocol) extends ChainableAction {
+                  protocol: SmtpProtocol) extends ChainableAction with Actor {
 
   val name = "sendMail"
 
   def execute(session: Session) {
-    val email: SimpleEmail = generateEmail(session)
-    val reqStart = System.currentTimeMillis()
-    val executionReport = sendMail(email)
-    val reqEnd = System.currentTimeMillis()
-
-    statsEngine.logResponse(session, requestName, ResponseTimings(reqStart, reqEnd), executionReport.status, None, executionReport.errorMessage)
+    def readSession(variableName: String): String = session.apply(variableName).as[String]
+    val request = SendMailRequest(session = session,
+      host = protocol.host,
+      port = protocol.port,
+      ssl = protocol.ssl,
+      from = readSession(UserFeeder.UsernameSessionParam),
+      to = readSession(UserFeeder.UsernameSessionParam),
+      subject = subject,
+      body = body,
+      credentials = readCredentials(readSession)(protocol))
+
+    context.actorOf(SmtpHandler.props()) ! request
 
     next ! session
   }
 
-  private def generateEmail(session: Session): SimpleEmail = {
-    def sessionVariable(variableName: String): String = {
-      session.apply(variableName).as[String]
-    }
-
-    val email = new SimpleEmail()
-    val userEmail = sessionVariable(UserFeeder.UsernameSessionParam)
-    if (protocol.auth) {
-      email.setAuthentication(sessionVariable(UserFeeder.UsernameSessionParam), sessionVariable(UserFeeder.PasswordSessionParam))
-    }
-    email.setHostName(protocol.host)
-    email.setSSLOnConnect(protocol.ssl)
-    email.setSmtpPort(protocol.port)
-    email.setSubject(subject + " [" + Math.random() + "]")
-    email.setMsg(body)
-    email.setFrom(userEmail)
-    email.addTo(userEmail)
-    email
+  override def receive: Receive = {
+    case session: Session => execute(session)
+    case executionReport: ExecutionReport =>
+      statsEngine.logResponse(executionReport.session, requestName, executionReport.responseTimings, executionReport.status, None, executionReport.errorMessage)
+    case Failure(e) =>
+      logger.error("Exception caught while sending mail", e)
```

Comment:
What do you mean by report? To who?

If the execution failed for any reason, the failure is handled by a failed execution report. Here the error is for our code.